src/hotspot/share/utilities/elfFile.cpp

```diff
@@ -1423,7 +1423,7 @@ bool DwarfFile::LineNumberProgram::apply_extended_opcode() {
         // Must be an unsigned integer as specified in section 6.2.2 of the DWARF 4 spec for the discriminator register.
         return false;
       }
-      _state->_discriminator = discriminator;
+      _state->_discriminator = static_cast<uint32_t>(discriminator);
```

Comment:
@chhagedorn These fields are declared as 32 bits but are read out of the dwarf info as 64 bits.  Should they be 64 bits?

Reply:
According to the DWARF 4 spec, the discriminator is an unsigned integer (i.e. 32 bits). However, `read_uleb128()` could either read 1, 2, 4 or 8 bytes and thus returns a 64 bit number. I guess it's fine to cast it like and keep the discriminator as 32 bit number as specified in the spec. We verify above that `read_uleb128()` did not read more than 4 bytes.

Reply:
According to the DWARF 4 spec, the discriminator is an unsigned integer (i.e. 32 bits). Also, the other fields like `isa or `column` are unsigned integer. I think we would be good to keep them as unsigned integer to follow the spec. 

`read_uleb128()` reads up to 8 bytes and just returns the result in a `uint64_t`. But sometimes, it should only read 1, 2, or 4 bytes. In that case, we additionally specify `check_size` to ensure that we can safely cast it. I guess we could change these calls to `read_uleb128()` to return a more precise type instead of relying on `check_size` passed in by a caller + casting the result in the caller.

Reply:
When we read the 64 bit values in the code, the values are 64 bits though.  So static_cast<> to narrow the result is more correct ?

Reply:
This code:
```
      uint64_t isa;
      if (!_reader.read_uleb128(&isa, 4)) {
        // isa register is 4 bytes wide.
        return false;
      }
      _state->_isa = isa;  // only save 4 bytes
```
returns 64 bit of information into isa from this read_uleb128 call, even though it passes 4.

Reply:
I've quickly skimmed through the usages of `read_uleb128`. We only seem to be reading either directly into a proper `uint64_t` or we are reading 4 bytes (i.e. `check_size = 4`). In the latter case, we could either:
- add a `static_cast` for the read value when storing it into a 32 bit field, as for example for `isa` above (i.e. `_state->_isa = static_cast<uint32_t>(isa)`).
- provide a 32 bit `read_uleb128()` version that hides the casting.

I guess it's fine to just go with a `static_cast` when storing to a field for the few cases we have.

Reply:
That looks good!